pkg/virtual/framework/dynamic/apiserver/openapi.go

```diff
@@ -0,0 +1,383 @@
+/*
+Copyright 2025 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"bytes"
+	"crypto/sha512"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+
+	"github.com/emicklei/go-restful/v3"
+	"github.com/go-logr/logr"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/endpoints/discovery"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/server/mux"
+	"k8s.io/klog/v2"
+	"k8s.io/kube-openapi/pkg/builder3"
+	"k8s.io/kube-openapi/pkg/common"
+	"k8s.io/kube-openapi/pkg/common/restfuladapter"
+	"k8s.io/kube-openapi/pkg/handler3"
+	"k8s.io/kube-openapi/pkg/spec3"
+	"k8s.io/utils/keymutex"
+	"k8s.io/utils/lru"
+
+	"github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/apidefinition"
+	dyncamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+	apisv1alpha1 "github.com/kcp-dev/kcp/sdk/apis/apis/v1alpha1"
+)
+
+const (
+	// DefaultServiceCacheSize is the default size of the OpenAPI service cache.
+	DefaultServiceCacheSize = 50
+)
+
+type openAPIServiceItem struct {
+	done    chan struct{}
```

Comment:
```suggestion
	done    <- chan struct{}
```
Makes clear who closes it.

Reply:
I tried that but it throws:

> invalid operation: cannot close receive-only channel item.done (variable of type <-chan struct{})compiler[InvalidClose](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#InvalidClose)

Reply:
Problem here is that you are doing `close(item.done)` later on I suppose. For @sttts' suggestion, I suppose you could close `doneCh`, not `item.done`, in the `defer` block in the later part of the code.